### PR TITLE
First test cases for HostBufferIntern

### PR DESCRIPTION
**Pull request by erikzenker**
This is how the test could look like. 
Both test are validated over DIM1, DIM2 and DIM3 and
over several number of elements per dimension.

### CODE REVIEW

**Comment by BenjaminW3** [comment]
You could possibly also use BOOST_AUTO_TEST_CASE_TEMPLATE so that the for_each is done by the library. This would enable you to correctly use per test fixtures.

**Comment by erikzenker** [comment]
Unfortunately, `BOOST_AUTO_TEST_CASE_TEMPLATE` does not compile for me with nvcc:

    /opt/pkg/compiler/gnu/gcc/4.9.2/include/c++/4.9.2/bits/basic_string.h(444): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_empty_rep_storage" is undefined in device code

    /opt/pkg/devel/boost/1.56.0/gnu/4.9.2/64/opt/include/boost/test/unit_test_suite_impl.hpp(376): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_terminal" is undefined in device code

    /opt/pkg/devel/boost/1.56.0/gnu/4.9.2/64/opt/include/boost/test/unit_test_suite_impl.hpp(377): error: identifier "_ZTIi" is undefined

    /opt/pkg/compiler/gnu/gcc/4.9.2/include/c++/4.9.2/bits/basic_string.h(444): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_empty_rep_storage" is undefined in device code

    /opt/pkg/devel/boost/1.56.0/gnu/4.9.2/64/opt/include/boost/test/unit_test_suite_impl.hpp(376): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_terminal" is undefined in device code

    /opt/pkg/devel/boost/1.56.0/gnu/4.9.2/64/opt/include/boost/test/unit_test_suite_impl.hpp(377): error: identifier "_ZTIl" is undefined

    /opt/pkg/compiler/gnu/gcc/4.9.2/include/c++/4.9.2/bits/basic_string.h(444): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_empty_rep_storage" is undefined in device code

    /opt/pkg/devel/boost/1.56.0/gnu/4.9.2/64/opt/include/boost/test/unit_test_suite_impl.hpp(376): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_terminal" is undefined in device code

    /opt/pkg/devel/boost/1.56.0/gnu/4.9.2/64/opt/include/boost/test/unit_test_suite_impl.hpp(377): error: identifier "_ZTIh" is undefined

That is the reasion why I wrote my own template test.

File: src/libPMacc/test/memory/HostBufferIntern/copyFrom.hpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+// #includes in "test/memoryUT.cu"
+
+/**
+ * Checks if data is copied correctly from device to
+ * host.
+ */
+struct CopyFromTest {
+
+    template<typename T_DIM>
```
Comment (psychocoderHPC) [inline src/libPMacc/test/memory/HostBufferIntern/copyFrom.hpp]: Could you please write `T_Dim` to full fill the [coding guide lines](https://github.com/ComputationalRadiationPhysics/picongpu/wiki/Coding-Guide-Lines)

File: src/libPMacc/test/memory/memoryUT.cu [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+};
+
+
+/**
+ * Defines for which numbers of elements a
+ * test should be verfied e.g. the size
+ * of a host or device buffer.
+ */
+template<typename T_DIM>
```
Comment (erikzenker) [inline src/libPMacc/test/memory/memoryUT.cu]: `T_Dim`

File: src/libPMacc/test/memory/HostBufferIntern/reset.hpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * Checks if the HostBufferIntern is reseted correctly to zero.
+ */
+
+
+
+
+struct ResetTest {
+
+    template<typename T_DIM>
```
Comment (erikzenker) [inline src/libPMacc/test/memory/HostBufferIntern/reset.hpp]: `T_Dim`

File: src/libPMacc/test/memory/HostBufferIntern/reset.hpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+// #includes in "test/memoryUT.cu"
+
+/**
+ * Checks if the HostBufferIntern is reseted correctly to zero.
+ */
```
Comment (psychocoderHPC) [inline src/libPMacc/test/memory/HostBufferIntern/reset.hpp]: emply lines

File: src/libPMacc/test/memory/memoryUT.cu [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+#include <boost/test/unit_test.hpp>
+#include <boost/mpl/list.hpp>
+#include <boost/mpl/for_each.hpp>
+
+// MPI
+#include <mpi.h> /* MPI_Init, MPI_Finalize */
+
+// PMacc
+#include <Environment.hpp>                       /* Environment */
+#include <dimensions/DataSpace.hpp>              /* DataSpace */
```
Comment (psychocoderHPC) [inline src/libPMacc/test/memory/memoryUT.cu]: Please remove the comment after the includes, both are equal.

File: src/libPMacc/test/memory/memoryUT.cu [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+ * of a host or device buffer.
+ */
+template<typename T_Dim>        
+std::vector<size_t> getElementsPerDim(){
+    std::vector<size_t> nElements;
+    std::vector<size_t> nElementsPerDim;
+
+    // Elements total
+    nElements.push_back(1);         // 1
+    nElements.push_back(1000);      // 1K
```
Comment (psychocoderHPC) [inline src/libPMacc/test/memory/memoryUT.cu]: Might be `1 * 1000` and `X * 1000` for the next lines

File: src/libPMacc/test/memory/memoryUT.cu [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+}
+    
+
+/**
+ * Definition of a list of dimension types. This
+ * List is used to test memory operations in
+ * each dimension setup automatically. For this
+ * purpose boost::mpl::for_each is used.
+ */
+struct Dim1_t { static int const value = DIM1; };
```
Comment (psychocoderHPC) [inline src/libPMacc/test/memory/memoryUT.cu]: Please use `boost::mpl::intergrals` 
e.g.
```
boost::mpl::intergral<int,value>
```

File: src/libPMacc/test/memory/memoryUT.cu [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+    // Elements total
+    nElements.push_back(1);
+    nElements.push_back(1 * 1000);
+    nElements.push_back(1 * 1000 * 1000);
+    nElements.push_back(1 * 1000 * 1000 * 10);
+
+    // Elements per dimension
+    for(unsigned i = 0; i < nElements.size(); ++i){
+        nElementsPerDim.push_back(std::pow(nElements[i], 1/static_cast<double>(T_Dim::value)));
```
Comment (psychocoderHPC) [inline src/libPMacc/test/memory/memoryUT.cu]: `1.0/static_cast<double>(T_Dim::value)`?

File: src/libPMacc/test/memory/HostBufferIntern/copyFrom.hpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    template<typename T_Dim>
+    void operator()(T_Dim){
+
+        typedef uint8_t Data;
+        typedef size_t  Extents;
+
+        std::vector<size_t> nElementsPerDim = getElementsPerDim<T_Dim>();
+        
+        for(unsigned i = 0; i < nElementsPerDim.size(); ++i){
+            ::PMacc::DataSpace<T_Dim::value> const dataSpace              = ::PMacc::DataSpace<T_Dim::value>::create(nElementsPerDim[i]);
```
Comment (psychocoderHPC) [inline src/libPMacc/test/memory/HostBufferIntern/copyFrom.hpp]: please remove the spaces before `=`

File: src/libPMacc/test/memory/memoryUT.cu [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    std::vector<size_t> nElementsPerDim;
+
+    // Elements total
+    nElements.push_back(1);
+    nElements.push_back(1 * 1000);
+    nElements.push_back(1 * 1000 * 1000);
+    nElements.push_back(1 * 1000 * 1000 * 10);
+
+    // Elements per dimension
+    for(unsigned i = 0; i < nElements.size(); ++i){
```
Comment (psychocoderHPC) [inline src/libPMacc/test/memory/memoryUT.cu]: `unsigned` -> `size_t` that `i` and `nElements.size()` had the some precision.

File: src/libPMacc/test/memory/HostBufferIntern/copyFrom.hpp [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+ * Checks if data is copied correctly from device to
+ * host.
+ */
+struct CopyFromTest {
+
+    template<typename T_Dim>
+    void operator()(T_Dim){
+
+        typedef uint8_t Data;
+        typedef size_t  Extents;
```
Comment (psychocoderHPC) [inline src/libPMacc/test/memory/HostBufferIntern/copyFrom.hpp]: please use only one space after the type

File: src/libPMacc/test/memory/memoryUT.cu (hunk 8 above)
Comment (erikzenker) [inline src/libPMacc/test/memory/memoryUT.cu]: The result should be a double, replaced by:
`static_cast<double>(1)/static_cast<double>(T_Dim::value)`

File: src/libPMacc/test/memory/HostBufferIntern/copyFrom.hpp (hunk 9 above)
Comment (erikzenker) [inline src/libPMacc/test/memory/HostBufferIntern/copyFrom.hpp]: These `=` are aligned for better readability, but I can also remove them.

File: src/libPMacc/test/memory/memoryUT.cu (hunk 10 above)
Comment (erikzenker) [inline src/libPMacc/test/memory/memoryUT.cu]: thx, well spotted

File: src/libPMacc/test/memory/HostBufferIntern/copyFrom.hpp (hunk 11 above)
Comment (erikzenker) [inline src/libPMacc/test/memory/HostBufferIntern/copyFrom.hpp]: Well, this aligned again for better readability.

File: src/libPMacc/test/memory/HostBufferIntern/copyFrom.hpp (hunk 9 above)
Comment (psychocoderHPC) [inline src/libPMacc/test/memory/HostBufferIntern/copyFrom.hpp]: Yes please remove

**Comment by erikzenker** [comment]
Update

**Comment by ax3l** [comment]
@erikzenker can you report the `BOOST_AUTO_TEST_CASE_TEMPLATE` in the [boost tracker](https://svn.boost.org/trac/boost)?

(Try again with latest CUDA 7.5 and latest boost before reporting a minimal example.)

**Comment by erikzenker** [comment]
I opened a [ticket](https://svn.boost.org/trac/boost/ticket/11822)

**Comment by ax3l** [comment]
thanks! :sparkles:

**Pull request merged**